src/Polly.Core/Utils/Pipeline/ComponentWithDisposeCallbacks.cs

```diff
@@ -0,0 +1,43 @@
+﻿namespace Polly.Utils.Pipeline;
+
+internal class ComponentWithDisposeCallbacks : PipelineComponent
+{
+    private readonly List<Action> _callbacks;
+
+    public ComponentWithDisposeCallbacks(PipelineComponent component, List<Action> callbacks)
+    {
+        Component = component;
+        _callbacks = callbacks.ToList();
+    }
+
+    internal PipelineComponent Component { get; }
+
+    public override void Dispose()
+    {
+        ExecuteCallbacks();
+
+        Component.Dispose();
+    }
+
+    public override ValueTask DisposeAsync()
+    {
+        ExecuteCallbacks();
+
+        return Component.DisposeAsync();
+    }
+
+    internal override ValueTask<Outcome<TResult>> ExecuteCore<TResult, TState>(
+        Func<ResilienceContext, TState, ValueTask<Outcome<TResult>>> callback,
+        ResilienceContext context,
+        TState state) => Component.ExecuteCore(callback, context, state);
+
+    private void ExecuteCallbacks()
+    {
+        foreach (var callback in _callbacks)
+        {
+            callback();
+        }
+
+        _callbacks.Clear();
```

Comment:
Do we need to worry about a race here if it gets disposed while still disposing?

Reply:
Not really, the pipeline is disposed only once and from a single thread.